### PR TITLE
libation: init at 11.1.0

### LINKED ISSUE #252519: Package request: Libation
**Project description**

This project is an open-source project to allow people to decrypt their audible libraries for use outside the audible app. It's currently only distributed via .deb and .rpm. 

**Metadata**

* homepage URL: https://github.com/rmcrackan/Libation
* source URL: https://github.com/rmcrackan/Libation/releases/tag/v11.0.2
* license: GPL-3.0
* platforms: linux

### PR DESCRIPTION
**Pull request by TomaSajt**
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/252519

This PR adds 1 package: `libation`
with 3 executables (`libation`, `libationcli`,`hangover`)

I couldn't try out all the features, but the basic ones seemed to work.

Not sure if I missed some dependencies.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/misc/libation/default.nix [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    libXi
+    libXcursor
+    libXext
+    libXrandr
+    fontconfig
+    glew
+
+    # For file dialog
+    gtk3
+    gdk-pixbuf
```
[the remark this answers is not in the document]
Comment (TomaSajt) [reply pkgs/applications/misc/libation/default.nix]: It turns out `gdk-pixbuf` wasn't really needed.
`gtk3` is used when displaying the file dialog. `gtk3`'s schemas are added already by the `wrapGAppsHook`, so this is only to add to the `LD_LIBRARY_PATH`. I've seen most other dotnet packages use `gtk3` in their `runtimeDeps`, and it worked here. If not specified, the program crashes when trying to open a file dialog.

**Review by mdarocha — APPROVED** [review]

File: pkgs/applications/misc/libation/default.nix [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  pname = "libation";
+  version = "11.0.3";
+
+  src = fetchFromGitHub {
+    owner = "rmcrackan";
+    repo = "Libation";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-PNsGIQ2h3r9y/m+3s9pmSQSXm6h5Z2GEvsgXSyeGvFs=";
+  };
+  sourceRoot = "source/Source";
```
Comment (pbsds) [inline pkgs/applications/misc/libation/default.nix:34]:
```suggestion
  sourceRoot = "${src.name}/Source";
```

File: pkgs/applications/misc/libation/default.nix [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+  preFixup = ''
+    makeWrapperArgs+=("''${gappsWrapperArgs[@]}")
+  '';
+
+  postFixup = ''
+    cd $out/bin
+    mv Libation libation
+    mv LibationCli libationcli
+    mv Hangover hangover
```
Comment (pbsds) [inline pkgs/applications/misc/libation/default.nix:96]:
```suggestion
    mv $out/bin/Libation $out/bin/libation
    mv $out/bin/LibationCli $out/bin/libationcli
    mv $out/bin/Hangover $out/bin/hangover
```

Comment (pbsds) [reply pkgs/applications/misc/libation/default.nix:96]: Seems the aarch64-darwin builder is running on a case-insensitive file system, the fix is perhaps to use an intermediate filename. If you do use that solution, note the reason why in a comment

**Review by pbsds — COMMENTED** [review]
Result of `nixpkgs-review pr 252999` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>libation</li>
  </ul>
</details>

Runs fine, but the build seems to produce a bunch of dlls and dylibs on linux. Is this a feature that could be disabled?

File: pkgs/applications/misc/libation/default.nix (hunk 3 above)
Comment (TomaSajt) [reply pkgs/applications/misc/libation/default.nix]: It now overrides the default dotnet fixup hook

**Comment by TomaSajt** [comment]
> Runs fine, but the build seems to produce a bunch of dlls and dylibs on linux. Is this a feature that could be disabled?

It turns out there's some custom removing logic in the original bundling script:
https://github.com/rmcrackan/Libation/blob/1702130b01de6c08ec5bd3f4010d81e71c6397fe/Scripts/Bundle_Debian.sh#L56

I'll se what I can do about it

**Comment by TomaSajt** [comment]
I have added some cleanup loosely based on the original script, using `stdenv.system` for what to keep

File: pkgs/by-name/li/libation/package.nix [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        --replace "/usr/bin/libation" "${meta.mainProgram}"
+  '';
+
+  # wrap manually, because we need lower case excutables
+  dontDotnetFixup = true;
+
+  preFixup = ''
+    pushd $out/lib/${pname}
+
+    rm -f *.x86.dll *.x64.dll
```
Comment (pbsds) [inline pkgs/by-name/li/libation/package.nix:73]:
```suggestion
    # remove binaries for other platform, like upstream does
    rm -f *.x86.dll *.x64.dll
```

File: pkgs/by-name/li/libation/package.nix [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        --replace "/usr/bin/libation" "${meta.mainProgram}"
+  '';
+
+  # wrap manually, because we need lower case excutables
+  dontDotnetFixup = true;
+
+  preFixup = ''
+    # remove binaries for other platform, like upstream does
+    pushd $out/lib/${pname}
+    rm -f *.x86.dll *.x64.dll
```
Comment (pbsds) [inline pkgs/by-name/li/libation/package.nix:76]:
```suggestion
    rm -f *.dll
```
Still a bunch of files left

Also, is the `-f` needed?

Comment (TomaSajt) [reply pkgs/by-name/li/libation/package.nix:76]: I don't know what I can do about those files...
Could you send me the remaining files, that should be temoved (that are not dotnet .dll files (those are different from normal windows dlls and are needed for all platforms))

About the -f, it looked like one of the possible library endings didn't actually appear (maybe arm64.so, i dont remember exactly)
But that may not be the case when building on other systems and I only have access to x64 linux. That's why -f is there: dont error if it doesnt exist

Comment (TomaSajt) [reply pkgs/by-name/li/libation/package.nix:76]: I have 3 normal `.so` files which do not have a prefix:
`libe_sqlite3.so` `libHarfBuzzSharp.so` and `libSkiaSharp.so`
would be great to know if they are `.dylib` files on darwin or not...
Other than this I think the current logic covers everything

Comment (pbsds) [reply pkgs/by-name/li/libation/package.nix:76]: > that are not dotnet .dll files (those are different from normal windows dlls and are needed for all platforms)

Ah, `file` tells me those are `PE32 executable (DLL)`, so I figured they were actual DLLs. Dotnet is weird.

I do not have darwin, so i cannot test that.

**Review by pbsds — APPROVED** [review]
Result of `nixpkgs-review pr 252999` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>libation</li>
  </ul>
</details>


- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package builds on x86_64-linux
- [x] executables tested on x86_64-linux
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.platforms` is set (handled by `buildDotnetModule`)
- [x] `meta.maintainers` is set
- [x] build time only dependencies are declared in `nativeBuildInputs`
- [x] source is fetched using the appropriate function
- [x] the list of `phases` is not overridden
- [x] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.
- [x] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed
- [x] patches that are remotely available are fetched rather than vendored

LGTM, thanks!

**Pull request merged**